src/testdrive/src/action.rs

```diff
@@ -57,6 +59,7 @@ pub struct Config {
 }
 
 pub struct State {
+    /// A random number to distinguish each TestDrive run
```

Comment:
I'm not a big fan of adding arbitrary docstrings to things like this. This is private API; I don't really think it needs to be documented, to be honest.

Reply:
Yeah, I'm sorry if this feels really nitpicky, but what I specifically find undesirable is having structs that are sporadically documented like this.  A struct with just a smattering of its fields documented just looks unfinished and wrong to my eye, and I find it very hard to see past that when I'm reading the code. For me, adding a one-off comment like this takes this struct from a polished and succinct representation of testdrive's state, though perhaps a somewhat underdocumented one, to one that is looks WIP, waiting for someone to come along and describe what the rest of the fields mean.

I just landed a PR I think you'll like that cleans up the documentation for testdrive's public API (https://github.com/MaterializeInc/materialize/pull/5308), so that mtrlz.dev/api/rust will render nicely. It adds a a complete set of comments to the `Config` struct.

To back up for a second, I'm unconvinced of the value of this sort of documentation for internal API. I think the idea that folks can pop into a codebase and have a complete set of accurate docstrings to show them around is unrealistic for any real-world code, and that the process of building a mental model of code is much more about spending time internalizing the various objects and how they fit together. In my experience, making things work more obviously is worth 100x more writing down how and why they behave non-obviously.

Anyway, I'm willing to put that opinion on hold, because I realize it is not widely popular with many other folks here. But my ask in return is that we are very intentional and precise when we do add documentation.

Specifically, if we're going to document `State`, here is how I would prefer for it to look!

```rust
/// The state associated with executing a single testdrive script.
///
/// A state is created at the beginning of each testdrive script execution,
/// then inspected and possibly mutated by each action in turn. The various
/// `reset` methods can be invoked at various points in the state's lifecycle
/// to clean up any external state modified by the script's actions.
pub struct State {
    /// A random number that uniquely identifies this state.
    ///
    /// Its value is exposed to scripts via the `testdrive.seed` variable. The
    /// seed is intended to be incorporated in the naming of objects stored in
    /// external state, so that multiple testdrive scripts can target the same
    /// external state simultaneously without conflicting. For example,
    /// `kafka::CreateTopicAction` incorporates the seed into the topic name.
    seed: u32,
    /// A temporary directory into which the script can emit data.
    ///
    /// The directory is destroyed when the state is dropped. The path of the
    /// directory is exposed to scripts via the `testdrive.temp_dir` variable.
    temp_dir: tempfile::TempDir,
    /// The optional path to the catalog backing the materialized instance to
    /// which testdrive connects.
    ///
    /// If present, actions can inspect the catalog contents as they like, to
    /// validate that the catalog matches their expectations.
    materialized_catalog_path: Option<PathBuf>,
    /// The address of the materialized instance to connect to.
    materialized_addr: String,
    /// A PostgreSQL client that is connected to the materialized instance
    /// specified by `materialized_addr`.
    pgclient: tokio_postgres::Client,
    /// The URL of a Confluent Schema Registry.
    ///
    /// The URL is exposed to scripts via the `testdrive.schema-registry-url`
    /// variable.
    schema_registry_url: Url,
    /// A schema registry client that is connected to the schema registry
    /// specified by `schema_registry_url`.
    ccsr_client: ccsr::Client,
    /// The address of a Kafka broker to connect to.
    ///
    /// The address is exposed to scripts via the `testdrive.kafka-addr`
    /// variable.
    kafka_addr: String,
    /// A Kafka admin client connected to the Kafka broker specified by `kafka_addr`.
    kafka_admin: rdkafka::admin::AdminClient<rdkafka::client::DefaultClientContext>,
    /// The options that should be used when making requests of the admin
    /// broker named in `kafka_admin`.
    kafka_admin_opts: rdkafka::admin::AdminOptions,
    /// The configuration options that should be used when creating Kafka
    /// clients that target the Kafka broker specified by `kafka_addr`.
    kafka_config: ClientConfig,
    /// A Kafka producer connected to the Kafka broker specified by
    /// `kafka_addr`.
    kafka_producer: rdkafka::producer::FutureProducer<rdkafka::client::DefaultClientContext>,
    /// A map from name to partition count for all Kafka topics created with
    /// this state.
    kafka_topics: HashMap<String, usize>,
    /// The AWS region to connect to.
    ///
    /// The name of the AWS region and the custom endpoint, if any, are exposed
    /// to the testdrive script as `testdrive.aws-region` and
    /// `testdrive.aws-endpoint`, respectively.
    aws_region: rusoto_core::Region,
    /// The account number for the account associated with AWS credentials
    /// specified in `aws_credentials`.
    aws_account: String,
    /// The credentials to use when connecting to AWS services.
    ///
    /// The access key ID, secret access key, and token within are exposed to
    /// scripts via the `testdrive.aws-access-key-id`,
    /// `testdrive.aws-secret-access-key`, and `testdrive.aws-token` variables,
    /// respectively.
    aws_credentials: AwsCredentials,
    /// A Kinesis client authenticated using `aws_credentials`.
    kinesis_client: KinesisClient,
    /// The names of the Kinesis streams created with this state.
    ///
    /// Calling [`State::reset_kinesis`] will attempt to delete these Kinesis
    /// streams.
    kinesis_stream_names: Vec<String>,
}
```

Reply:
Brandon and I talked offline; I'm going to periodically swoop in and use the comments Brandon adds to determine where things are underdocumented and add more cohesive docs. Therefore this is fine for now!